// discord.js-commando/src/commands/util/ping.js
const oneLine = require('common-tags').oneLine;
const Command = require('../base');

module.exports = class PingCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'ping',
			group: 'util',
			memberName: 'ping',
			description: 'Vérifie le ping du bot sur le serveur.',
			throttling: {
				usages: 5,
				duration: 10
			}
		});
	}

	async run(msg) {
		if(!msg.editable) {
			const pingMsg = await msg.reply('Pinging...');
			return pingMsg.edit(oneLine`
				${msg.channel.type !== 'dm' ? `${msg.author},` : ''}
				Pong! Le retour a pris ${pingMsg.createdTimestamp - msg.createdTimestamp}ms.
				${this.client.ping ? `Le rythme du battement de coeur est de ${Math.round(this.client.ping)}ms.` : ''}
			`);
		} else {
			await msg.edit('Pinging...');
			return msg.edit(oneLine`
				Pong! Le retour a pris ${msg.editedTimestamp - msg.createdTimestamp}ms.
				${this.client.ping ? `Le rythme du battement de coeur est de ${Math.round(this.client.ping)}ms.` : ''}
			`);
		}
	}
};
